### PR TITLE
initial commit of Pipeline support

### PR DESCRIPTION
**Pull request by gkorland**

### CODE REVIEW

File: src/main/java/com/redislabs/redisgraph/RedisGraphPipeline.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+     * Synchronize pipeline by reading all responses. This operation close the pipeline. Whenever
+     * possible try to avoid using this version and use Pipeline.sync() as it won't go through all the
+     * responses and generate the right response type (usually it is a waste of time).
+     * @return A list of all the responses in the order you executed them.
+     */
+    List<Object> syncAndReturnAll();
+    
+    /**
+     * Synchronize pipeline by reading all responses. This operation close the pipeline. In order to
+     * get return values from pipelined commands, capture the different Response&lt;?&gt; of the
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/RedisGraphPipeline.java:154]: format of `Response&lt;?&gt;`

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    /**
+     * Execute a Cypher query.
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @return a response which builds the result set with the query answer.
+     */
+    @Override
+    public Response<ResultSet> query(String graphId, String query) {
+        client.sendCommand(RedisGraphCommand.QUERY, graphId, query, "--COMPACT");
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:39]: don't we have a static string for `--COMPACT`?

File: src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+     *
+     * NOTE: timeout is simply sent to DB. Socket timeout will not be changed.
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @param timeout
+     * @return a response which builds the result set with the query answer.
+     */
+    @Override
+    public Response<ResultSet> query(String graphId, String query, long timeout) {
+        client.sendCommand(RedisGraphCommand.QUERY, graphId, query, "--COMPACT", "TIMEOUT", Long.toString(timeout));
```
Comment (DvirDukhan) [inline src/main/java/com/redislabs/redisgraph/impl/api/RedisGraphPipeline.java:76]: don't we have static strings for compact and timeout?

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+    }
+
+    @After
+    public void deleteGraph() {
+        api.deleteGraph("social");
+        api.close();
+    }
+
+    @Test
+    public void testCreateNode() {
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:34]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+
+        try {
+            resultSet.next();
+            Assert.fail();
+        } catch (NoSuchElementException ignored) {
+        }
+    }
+
+    @Test
+    public void testCreateLabeledNode() {
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:56]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+        // Create a node with a label
+        ResultSet resultSet = api.query("social", "CREATE (:human{name:'danny',age:12})");
+        Assert.assertFalse(resultSet.hasNext());
+        Assert.assertEquals("1", resultSet.getStatistics().getStringValue(Label.NODES_CREATED));
+        Assert.assertEquals("2", resultSet.getStatistics().getStringValue(Label.PROPERTIES_SET));
+        Assert.assertNotNull(resultSet.getStatistics().getStringValue(Label.QUERY_INTERNAL_EXECUTION_TIME));
+    }
+
+    @Test
+    public void testConnectNodes() {
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:66]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertFalse(resultSet.hasNext());
+        Assert.assertNull(resultSet.getStatistics().getStringValue(Label.NODES_CREATED));
+        Assert.assertNull(resultSet.getStatistics().getStringValue(Label.PROPERTIES_SET));
+        Assert.assertEquals(1, resultSet.getStatistics().relationshipsCreated());
+        Assert.assertEquals(0, resultSet.getStatistics().relationshipsDeleted());
+        Assert.assertNotNull(resultSet.getStatistics().getStringValue(Label.QUERY_INTERNAL_EXECUTION_TIME));
+    }
+
+    @Test
+    public void testDeleteNodes(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:83]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertNull(deleteResult.getStatistics().getStringValue(Label.RELATIONSHIPS_CREATED));
+        Assert.assertEquals(1, deleteResult.getStatistics().relationshipsDeleted());
+        Assert.assertEquals(1, deleteResult.getStatistics().nodesDeleted());
+
+        Assert.assertNotNull(deleteResult.getStatistics().getStringValue(Label.QUERY_INTERNAL_EXECUTION_TIME));
+
+    }
+
+    @Test
+    public void testDeleteRelationship(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:114]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertNull(deleteResult.getStatistics().getStringValue(Label.NODES_DELETED));
+        Assert.assertEquals(1, deleteResult.getStatistics().relationshipsDeleted());
+
+        Assert.assertNotNull(deleteResult.getStatistics().getStringValue(Label.QUERY_INTERNAL_EXECUTION_TIME));
+
+    }
+
+
+    @Test
+    public void testIndex() {
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:135]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+
+        ResultSet deleteExistingIndexResult = api.query("social", "DROP INDEX ON :person(age)");
+        Assert.assertFalse(deleteExistingIndexResult.hasNext());
+        Assert.assertNotNull(deleteExistingIndexResult.getStatistics().getStringValue(Label.INDICES_DELETED));
+        Assert.assertEquals(1, deleteExistingIndexResult.getStatistics().indicesDeleted());
+
+    }
+
+    @Test
+    public void testHeader(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:162]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertNotNull(schemaNames);
+        Assert.assertEquals(3, schemaNames.size());
+        Assert.assertEquals("a", schemaNames.get(0));
+        Assert.assertEquals("r", schemaNames.get(1));
+        Assert.assertEquals("a.age", schemaNames.get(2));
+
+    }
+
+    @Test
+    public void testRecord(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:188]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertEquals( 32L, ((Long)record.getValue(3)).longValue());
+        Assert.assertEquals( 32L, ((Long)record.getValue("a.age")).longValue());
+        Assert.assertEquals( "roi", record.getString("a.name"));
+        Assert.assertEquals( "32", record.getString("a.age"));
+
+    }
+
+    @Ignore
+    @Test
+    public void tinyTestMultiThread(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:305]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+        for (int i =0; i < 10000; i++){
+            List<ResultSet> resultSets = IntStream.range(0,16).parallel().
+                    mapToObj(
+                            j-> api.query("social", "MATCH (a:person) RETURN a")).
+                    collect(Collectors.toList());
+        }
+    }
+
+    @Test
+    public void testMultiThread(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:317]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+            Record record = resultSet.next();
+            Assert.assertFalse(resultSet.hasNext());
+            Assert.assertEquals(Arrays.asList("a", "r"), record.keys());
+            Assert.assertEquals(Arrays.asList(expectedNode, expectedEdge), record.values());
+        }
+    }
+
+
+    @Test
+    public void testAdditionToProcedures(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:401]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 15]
```diff
@@ -0,0 +1,642 @@
...
+        Assert.assertTrue(resultSet.hasNext());
+        record = resultSet.next();
+        Assert.assertFalse(resultSet.hasNext());
+        Assert.assertEquals(Arrays.asList("a", "r"), record.keys());
+        Assert.assertEquals(Arrays.asList(expectedNode, expectedEdge), record.values());
+
+    }
+
+    @Test
+    public void testEscapedQuery() {
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:474]: not relevant here

File: src/test/java/com/redislabs/redisgraph/PipelineTest.java [hunk 16]
```diff
@@ -0,0 +1,642 @@
...
+            Assert.assertTrue(resultSet.hasNext());
+            record = resultSet.next();
+            Assert.assertFalse(resultSet.hasNext());
+            Assert.assertEquals(Arrays.asList("label"), record.keys());
+            Assert.assertEquals("Person", record.getValue("label"));
+        }
+    }
+
+    @Test
+    public void testMultiExecWithReadOnlyQueries(){
```
Comment (DvirDukhan) [inline src/test/java/com/redislabs/redisgraph/PipelineTest.java:569]:
```suggestion
    public void testPipelineWithReadOnlyQueries(){
```

**Review by DvirDukhan — CHANGES_REQUESTED** [review]

File: src/main/java/com/redislabs/redisgraph/RedisGraphPipeline.java (hunk 1 above)
Comment (gkorland) [reply src/main/java/com/redislabs/redisgraph/RedisGraphPipeline.java:154]: `Response<?>`

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**